core/iwasm/common/wasm_runtime_common.c

```diff
@@ -1136,6 +1135,38 @@ wasm_runtime_load(uint8 *buf, uint32 size, char *error_buf,
         set_error_buf(error_buf, error_buf_size,
                       "WASM module load failed: magic header not detected");
     return NULL;
+
+register_module:
+    module_common = register_module_with_null_name(module_common, error_buf,
+                                                   error_buf_size);
+    if (module_common == NULL) {
+        return NULL;
+    }
+
+#if WASM_ENABLE_SHARED_MEMORY != 0
+    module = (WASMModule *)module_common;
+    if (os_mutex_init(&module->mem_lock) != 0) {
+        set_error_buf(error_buf, error_buf_size,
+                      "WASM module load failed: unable to init memory mutex");
```

Comment:
Don't initialize `module->mem_lock` here, since the `module_common` here may be aot module (AOTModule), which hasn't field mem_lock, mem_lock is only used in interpreter mode.
Had better init it in `create_module` in wasm_loader.c/wasm_mini_loader.c instead:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_loader.c#L3877
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_mini_loader.c#L2736
And destroy it in wasm_loader_unload:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_loader.c#L4249
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_mini_loader.c#L2942

But another solution is to reuse the main instance's mem_lock for child instance while instantiating the child instance in thread creating, similar to the wasi_ctx (note the `is_sub_inst` flag, false for main instance, true means child instance created by new thread):
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_runtime.c#L1954-L1966
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_runtime.c#L2127

My personal preference is to use the latter method, had better isolate the instance data as long as we can, so two different instances (created by wasm_runtime_instantiate but not pthread_create callback) won't share the same mem_lock.

Reply:
Thanks for the review. For the second approach I need to check, the problem is that (at least for WASI threads) each thread uses a separate module instance and the lock should be shared.

Reply:
I went for the first approach.
PS: I'm working on a separate PR to fix some other race conditions I found while testing.

Reply:
I agree with @wenyongh's second approach. The module may be instantiated multiple times (not a thread of main instance, directly instantiate from a module), then these instances are independent and should not rely on the same lock.
If the lock is defined in module, then atomic operation in one instance will influence another independence instance.

Reply:
I checked the code, seems in current implementation the shared memory is shared globally (all instances from all clusters share the same memory). But I checked the proposal again, seems it should be better to only share the memory inside a cluster.
If we want to modify the sharing scope to cluster later, had better not put the lock inside module

Reply:
I opened an issue to track the issue about memory is shared globally: https://github.com/bytecodealliance/wasm-micro-runtime/issues/1962

Reply:
I understand what you say, if a module is instantiated multiple times, the lock (and with the current implementation the memory too https://github.com/bytecodealliance/wasm-micro-runtime/blob/1c17665f68e118cd9c6bc9b8782f1d0cda5994c9/core/iwasm/interpreter/wasm_runtime.h#L265) would be shared as well. That's good for threads but not in the general case.

I'll see if I can modify this PR to handle that.

Reply:
Actually, what I can do in this PR is to just move the lock to `WASMMemoryInstance` instead of `WASMModule`. That should be compatible with what we want to do in #1962. What do you think?

Reply:
> Actually, what I can do in this PR is to just move the lock to `WASMMemoryInstance` instead of `WASMModule`. That should be compatible with what we want to do in #1962. What do you think? @xujuntwt95329

Yes, sounds reasonable~

Reply:
After the suggestion from @wenyongh, I put it in `WASMSharedMemory`. We should be good to merge.